### PR TITLE
Use RList for cmd_descriptors and rename cfg.newtab to cfg.comphint

### PR DESCRIPTION
**Pull request by MaskRay**

### CODE REVIEW

File: libr/core/cmd.c [hunk 1]
```diff
@@ -27,32 +27,48 @@
...
+		RCmdDescriptor *d = R_NEW0 (RCmdDescriptor);				\
+		if (d) {																						\
+			d->name = #name_;																	\
+			d->help_msg = help_msg_##name_;										\
+			d->help_detail = help_detail_##name_;							\
+			r_list_append (core->cmd_descriptors, d);					\
+		}																										\
+	}
+
+#define DEFINE_CMD_DESCRIPTOR_WITH_DETAIL2(core, name_)	\
```
Comment (radare) [inline libr/core/cmd.c]: backslash should  be only 1 char after the last char of the line, not at random places

File: libr/core/cmd.c [hunk 2]
```diff
@@ -3111,31 +3131,30 @@ static int compare_cmd_descriptor_name(const void *a, const void *b) {
...
+					RCmdDescriptor *d = R_NEW0 (RCmdDescriptor);
+					r_list_append (core->cmd_descriptors, d);
+					x->sub[*p] = d;
+				} else {
+					x->sub[*p] = y;
 				}
-				x->sub[*p] = &cmd_descriptors[cmd_descriptors_len++];
+			} else if (!p[1]) {
+				eprintf ("Command '%s' is duplicated, please check\n", y->name);
+				exit (1);
```
Comment (radare) [inline libr/core/cmd.c]: do not exit EVER

File: libr/core/cmd.c [hunk 3]
```diff
@@ -3111,31 +3131,30 @@ static int compare_cmd_descriptor_name(const void *a, const void *b) {
...
-		const ut8 *p;
-		RCmdDescriptor *x = &core->cmd_descriptor;
-		for (p = (const ut8*)cmd_descriptors[i].name; *p; p++) {
+	const ut8 *p;
+	RListIter *iter;
+	RCmdDescriptor *x, *y;
+	int n = core->cmd_descriptors->length;
+	r_list_sort (core->cmd_descriptors, compare_cmd_descriptor_name);
+	r_list_foreach (core->cmd_descriptors, iter, y) {
+		if (--n < 0) break;
```
Comment (radare) [inline libr/core/cmd.c]: missing {} braces

File: libr/core/cmd.c [hunk 4]
```diff
@@ -3111,31 +3131,30 @@ static int compare_cmd_descriptor_name(const void *a, const void *b) {
...
+		x = &core->root_cmd_descriptor;
+		for (p = (const ut8 *)y->name; *p; p++) {
 			if (!x->sub[*p]) {
-				if (cmd_descriptors_len >= MAX_NUM_CMD_DESCRIPTORS) {
-					eprintf ("Please increase MAX_NUM_CMD_DESCRIPTORS in libr/core/cmd.c\n");
-					exit(1);
+				if (p[1]) {
+					RCmdDescriptor *d = R_NEW0 (RCmdDescriptor);
```
Comment (radare) [inline libr/core/cmd.c]: check null ptr

File: libr/core/cmd.c [hunk 5]
```diff
@@ -3199,13 +3205,35 @@ R_API void r_core_cmd_init(RCore *core) {
...
 	r_cmd_add (core->rcmd, "/",        "search kw, pattern aes", &cmd_search);
-	cmd_search_init ();
 	r_cmd_add (core->rcmd, "-",        "open cfg.editor and run script", &cmd_stdin);
 	r_cmd_add (core->rcmd, "(",        "macro", &cmd_macro);
 	r_cmd_add (core->rcmd, "u",        "uname/undo", &cmd_uname);
 	r_cmd_add (core->rcmd, "quit",     "exit program session", &cmd_quit);
 	r_cmd_add (core->rcmd, "Q",        "alias for q!", &cmd_Quit);
 	r_cmd_add (core->rcmd, "L",        "manage dynamically loaded plugins", &cmd_plugins);
 
+	cmd_search_init (core);
```
Comment (radare) [inline libr/core/cmd.c]: make a single function that calls all those things, but ideally it should be defined in an array of structs with { "quit", "exit program", cmd_exit_init }, ...

**Comment by radare** [comment]
i like the direction where things are going, but not with spaguetti code. all those commands of commands should be defined statically and then use a for loop to construct them all instead of calling all several functions

File: libr/core/cmd.c [hunk 6]
```diff
@@ -27,32 +27,48 @@
...
-	REGISTER_CMD_DESCRIPTOR(name_);																				\
-	cmd_descriptors[cmd_descriptors_len-1].name = #name_;									\
-	cmd_descriptors[cmd_descriptors_len-1].help_msg = help_msg_##name_;		\
-	cmd_descriptors[cmd_descriptors_len-1].help_detail = help_detail_##name_;
-
-#define DEFINE_CMD_DESCRIPTOR_SPECIAL(cmd, name_)                          \
-	REGISTER_CMD_DESCRIPTOR(name_);                                     \
-	cmd_descriptors[cmd_descriptors_len-1].name = #cmd;                 \
-	cmd_descriptors[cmd_descriptors_len-1].help_msg = help_msg_##name_; \
+#define DEFINE_CMD_DESCRIPTOR(core, name_)				\
```
Comment (radare) [inline libr/core/cmd.c]: it is not recommended to use large macros, better move as much logic as possible in a separate static function

File: libr/core/cmd.c (hunk 1 above)
Comment (MaskRay) [inline libr/core/cmd.c]: Emacs did the indentation for me. What `set ts` do you use?

File: libr/core/cmd.c (hunk 4 above)
Comment (MaskRay) [inline libr/core/cmd.c]: `r_list_append` does not append the element if `data` if NULL. So, the null case is already handled.

File: libr/core/cmd.c (hunk 5 above)
Comment (MaskRay) [inline libr/core/cmd.c]: Can I add another function `r_cmd_add2` which adds the `cmd_*_init` argument to `r_cmd_add`?  Can you give it a name?

File: libr/core/cmd.c (hunk 3 above)
Comment (MaskRay) [inline libr/core/cmd.c]: Done

File: libr/core/cmd.c (hunk 5 above)
Comment (radare) [inline libr/core/cmd.c]: no, the best thing would be to accept an array of structs.

typedef struct {
  const char *name;
  const char *desc;
  RCmdCallback cb;
  RCmdInit init;
} RCmdEntry;

RCmdEntry entries[] = {
 { "quit", "exit program", cmd_quit, cmd_exit_init },
 { ...} 
 {NULL}
}

r_cmd_add_all (core->cmd, entries);

**Comment by radare** [comment]
let's merge for now, we should unroll that spaguetti later

**Pull request merged**